### ISSUE TITLE
Support `Include` in ssh config

### ISSUE THREAD

**Issue by YDX-2147483647** [microsoft/vscode-remote-release#9668]
In OpenSSH, you can use [`Include`](https://man.openbsd.org/ssh_config#Include) to split ssh config into multiple files. 

For example:

`~/.ssh/config`:

```
Include config.d/*
```

`~/.ssh/config.d/company`: (copied from #9370)

```
Host ServerA
    HostName website.domain.com
    User ubuntu
    IdentityFile {myKey}
```

## Current

Remote - SSH extension ignores `Include`.

![](https://github.com/microsoft/vscode-remote-release/assets/73375426/c5289fd2-f864-4f2c-a668-18fc85299130)

## Desired result

It should be equivalent to the following.

![](https://github.com/microsoft/vscode-remote-release/assets/73375426/1554c45e-8fa2-4d76-854d-ed7de40ed949)

**Comment by roblourens** [comment]
This should work, it's working for me on mac, I'll check Windows

**Comment by YDX-2147483647** [comment]
> I'll check Windows

Hi! Have you checked? Do you need any additional information?

**Comment by florentcpt** [comment]
Hello,

We have the same regression here (using Windows 11) with recent versions of the extension.

The problem is really coming from the wildcard, based on our tests:

```sh
# ~/.ssh/config
Include conf.d/*  # <-- Doesn't work (but used to)
Include conf.d/myconf.conf  # <-- Actually works
```

**Comment by YDX-2147483647** [comment]
> the wildcard

I can verify it on Windows 10. So currently a workaround would be expanding the wildcard on our own.

**Comment by Tony-Sol** [comment]
Not working on macos. Neither with nor without wildcard in `Include` section.
Here is example
<img width="1684" alt="w-wildcard" src="https://github.com/microsoft/vscode-remote-release/assets/33671815/2f2ce06c-2834-4172-a9b7-b24252b55685">
<img width="1684" alt="wo-wildcard" src="https://github.com/microsoft/vscode-remote-release/assets/33671815/94f4f9b7-6064-4fe4-ac44-d32c97ae25ff">

Also, i guess `Match all` is not a valid host to display in connections list

**Comment by den-is** [comment]
I confirm wildcard Include do not work on macos.
I have this in the of my ssh_config
```
Match all
Include ~/.ssh/config.d/*.config
```

VSCode is showing `all` in the list of available connections. `Match` and `Match all` should definitely be ignored by VSCode

**Comment by torchsmith** [comment]
Not working for me either on macos.

**Comment by torchsmith** [comment]
@roblourens Hi Rob, since this is standard OpenSSH functionality, do you think we can get some priority assigned to this issue?

**Comment by roblourens** [comment]
This code hasn't changed in quite a while. Can someone try downgrading the extension (in the extension view, right click > Install Another Version) to help me figure out when it changed? Otherwise I just need to add a lot of logging or something

**Comment by YDX-2147483647** [comment]
> Can someone try downgrading the extension

Here you are:

- 0.107.1 works,
- 0.108.0 does not.

Tested on Window 10, with `Include config.d/*` as `~/.ssh/config`.

**Comment by Tony-Sol** [comment]
>  someone try downgrading the extension

tested on macos 14.4.1

`Include config.d/*` not working if this sections is under `Match` group.

If use `Include` without any match group, it works on latest, otherwise - not working in any version.

Also, nested `Include`, like 

~/.ssh/config
```sshconfig
Include config.d/*
```
~/.ssh/config.d/config
```sshconfig
Include config.d/test/*
```
~/.ssh/config.d/test/config
```sshconfig
Host A
```

Works only since 0.66.0

**Comment by torchsmith** [comment]
Im on version v0.110.1 and it's not working with just the Include.

**Comment by torchsmith** [comment]
v0.107.1 does not work for me. macos.

`Include ~/.ssh/configs/conf.conf`

**Comment by torchsmith** [comment]
Agh, it works when I add "Match" before my Include statement. Just not without Match. Is this an error on my part?

**Comment by torchsmith** [comment]
New updates.

Latest version (v0.110.1) works with just Include.

The issue is that in order for it to work it needs to be at the the very top of my config. Before it was underneath a Host * statement. (Which works with normal ssh in terminal).

**Comment by roblourens** [comment]
I see, I think what is being described is essentially https://github.com/microsoft/vscode-remote-release/issues/4339. Will take a look when I'm able

**Comment by roblourens** [comment]
No, there's just a lot going on here #4339 still exists. But also it seems like glob includes regressed on Windows. Probably when I updated the 'glob' package. Taking a look.

**Comment by roblourens** [comment]
The 'glob' package behavior changed on Windows- fix coming in June. Sorry for the break.

**Comment by torchsmith** [comment]
@roblourens when will this update be pushed to stable so we can test it out?

**Comment by roblourens** [comment]
All VS Code/extension releases are paused right now for stabilization before the release. This will be in the next Remote-SSH pre-release towards the end of next week.

**Comment by Tony-Sol** [comment]
Is there issue about `Match` includes in servers list? Didn't find any :c

**Comment by den-is** [comment]
Any updates? When fix is going to be released?

**Comment by roblourens** [comment]
Sorry, the release got delayed due to technical difficulties 🙂 we are working on getting it out

**Comment by PillarsZhang** [comment]
Wow, after the update, the long-missed wildcard SSH hosts are back. Thank you so much! 😊

**Comment by tony-sol** [comment]
There is something wrong in here, i believe

<img width="801" alt="include-after-match" src="https://github.com/microsoft/vscode-remote-release/assets/33671815/e73bc33d-58bb-4b68-8074-7d78d110e517">
<img width="801" alt="include-before-match" src="https://github.com/microsoft/vscode-remote-release/assets/33671815/1bc5e414-1cef-43e2-bf73-6b7b83cd0928">

**Comment by den-is** [comment]
> Wow, after the update, the long-missed wildcard SSH hosts are back. Thank you so much!

@PillarsZhang what version are you talking about?
I still have original v0.112.0 from 2024-06-14, 08:44:49 which is not rendering my .ssh/config

**Comment by PillarsZhang** [comment]
@den-is I'm using `v0.113.2024061415 (pre-release)`, but when I reverted to the released version `v0.112.0`, all the hosts were still there. I'm not sure if it's because of an updated dependency.
I have VS Code and VS Code Insiders installed on two different computers, and both exhibit the same behavior.

**Comment by den-is** [comment]
any updates on this?
I thought you were going to release the last month. Now I suppose it is blocked by the another new VSCode version?
